src/metabase/models/card.clj

```diff
@@ -763,14 +795,43 @@
        ;; If we did have to generate the hashed entity_id, include it on the returned card as well.
        @hashed-eid (assoc :entity_id @entity-id)))))
 
+(defn- upgrade-card-schema-to-latest [card]
+  (if (and (:id card)
+           (or (:dataset_query card)
+               (:result_metadata card)
+               (:database_id card)
+               (:type card)))
+    ;; A plausible select to run the after-select logic on.
+    (if-not (:card_schema card)
+      ;; Plausible but no :card_schema - error.
+      (throw (ex-info "Cannot SELECT a Card without including :card_schema"
+                      {:card-id (:id card)}))
+      ;; Plausible and has the schema, so run the upgrades over it.
+      (loop [card card]
+        (if (= (:card_schema card) current-schema-version)
+          card
+          (let [new-version (inc (:card_schema card))]
+            (recur (assoc (upgrade-card-schema-to card new-version)
+                          :card_schema new-version))))))
```

Comment:
Should the `upgrade-card-schema-to` methods handle updating the card's `:card_schema`?

I guess these are all private, so maybe `upgrade-card-schema-to-latest` is the only acceptable entry point, but it feels strange to call `(upgrade-card-schema-to card N)` and get back a card with the old schema number.

Reply:
I don't want to make that a forget-able step. So the outer loop over the schema versions will either contain (1) a check that your updater did it right, or (2) code to just set it. The latter seemed more straightforward to me.